flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/JdbcSourceSplitReader.java

```diff
@@ -0,0 +1,291 @@
+                            "The SplitChange type of %s is not supported.",
+                            splitsChanges.getClass()));
+        }
+
+        LOG.debug("Handling split change {}", splitsChanges);
+        splits.addAll(splitsChanges.splits());
+    }
+
+    @Override
+    public void wakeUp() {}
+
+    @Override
+    public void close() throws Exception {
+        closeResultSetAndStatement();
+        if (connection != null && !connection.isClosed()) {
+            connection.close();
+        }
+        connection = null;
+        currentSplit = null;
+    }
+
+    @Override
+    public TypeInformation<T> getProducedType() {
+        return typeInformation;
+    }
+
+    private void checkSplitOrStartNext() {
+
+        try {
+            if (hasNextRecordCurrentSplit && resultSet != null) {
+                return;
+            }
+
+            final JdbcSourceSplit nextSplit = splits.poll();
+            if (nextSplit == null) {
+                throw new IOException("Cannot fetch from another split - no split remaining");
+            }
+            currentSplit = nextSplit;
+            openResultSetForSplit(currentSplit);
+        } catch (SQLException | ClassNotFoundException | IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void discardSplit(JdbcSourceSplit split) throws SQLException {
+        if (split.getOffset() != 0) {
+            hasNextRecordCurrentSplit = false;
+            currentSplitOffset = 0;
+            if (resultSet != null && !resultSet.isClosed()) {
+                resultSet.close();
+            }
+            if (statement != null && !statement.isClosed()) {
+                statement.close();
+            }
+            resultSet = null;
+            statement = null;
+            currentSplit = null;
+        }
+    }
+
+    private void getOrEstablishConnection() throws SQLException, ClassNotFoundException {
+        connection = connectionProvider.getOrEstablishConnection();
+        if (autoCommit == null) {
+            return;
+        }
+        if (autoCommit != connection.getAutoCommit()) {
+            connection.setAutoCommit(autoCommit);
+        }
+    }
+
+    private void openResultSetForSplit(JdbcSourceSplit split)
+            throws SQLException, ClassNotFoundException {
+        getOrEstablishConnection();
+        statement =
+                connection.prepareStatement(
+                        split.getSqlTemplate(), resultSetType, resultSetConcurrency);
+        if (split.getParameters() != null) {
+            Object[] objs = split.getParameters();
+            for (int i = 0; i < objs.length; i++) {
+                statement.setObject(i + 1, objs[i]);
+            }
+        }
+        statement.setFetchSize(resultSetFetchSize);
+        resultSet = statement.executeQuery();
+        hasNextRecordCurrentSplit = resultSet.next();
+        // TODO fine-grained EXACTLY_ONCE/AT_MOST_ONCE/AT_LEAST_ONCE will be introduced in the
```

Comment:
I'm not sure having such TODO will help
probably it's better to have a dedicated jira issue and mention it here
At least we could track it then

Reply:
deleted.